cosmo_tester/framework/config.py

```diff
@@ -0,0 +1,277 @@
+        # Currently, if we can load it then it's valid.
+        return True
+
+    def _generate_config(self, schema=None, raw_config=None):
+        schema = schema or self.schema
+        raw_config = raw_config or self.raw_config
+
+        # Get all namespaces and config entries at the current level
+        namespaces = [
+            key for key in schema.keys()
+            if key != '.is_namespace'
+            and schema[key].get('.is_namespace', False)
+        ]
+        config = {
+            k: v.get('default', NotSet)
+            for k, v in schema.items()
+            if k != '.is_namespace'
+        }
+
+        # Populate config values based on user configuration
+        for key, value in raw_config.items():
+            if key in config:
+                config[key] = value
+
+        # Populate namespaces
+        for namespace in namespaces:
+            config[namespace] = self._generate_config(
+                schema=schema[namespace],
+                raw_config=raw_config.get(namespace, {}),
+            )
+
+        return config
+
+    def __getitem__(self, item):
+        if not self._config:
+            self._config = self._generate_config()
+        config = self._config
+
+        if item in config.keys():
+            return config[item]
+        else:
+            if item in self.raw_config.keys():
+                raise KeyError(
+                    'Config entry {key} was supplied but was not in the '
+                    'schema. Please update the schema to use this config '
+                    'entry.'.format(key=item)
+                )
+            else:
+                raise KeyError(
+                    'Config entry {key} was not supplied and is not in '
+                    'schema.'.format(key=item)
+                )
+
+    def keys(self):
+        return self._generate_config().keys()
+
+    def values(self):
+        return self._generate_config().values()
+
+    def items(self):
+        return self._generate_config().items()
+
+
+def find_schemas():
+    # I'm not entirely keen on ignoring the userwarning here,
+    # but it seems like both Ubuntu and centos default to group
+    # writeable ~/.python-eggs, which triggers this warning.
+    with warnings.catch_warnings():
+        warnings.simplefilter('ignore', category=UserWarning)
+        schemas = pkg_resources.resource_listdir(
+            'cosmo_tester',
+            'config_schemas',
+        )
+        schemas = [
+            pkg_resources.resource_filename(
+                'cosmo_tester',
+                os.path.join('config_schemas', schema),
+            )
+            for schema in schemas
+        ]
+
+    return schemas
+
+
+def load_config(logger, config_file=None, missing_config_fail=True):
+    """Load the configuration from the specified file.
+    missing_config_fail determines whether the file being absent is fatal.
+    """
+    schemas = find_schemas()
+    config = None
+    try:
+        config = Config(
```

Comment:
It was deliberately done as a class so that it could validate that anything used in the tests is defined in the schema, avoiding silliness like having people inject extra stuff at runtime that's then effectively undocumented and unknowable for anyone who didn't realise that some out of the way fixture was mucking about.
The use of the schema is generally simple dict* and I'll add some stuff to the README once it's all finalised (and do a demo on it).
*- with the exception of .platform, which was deliberately introduced to get the current platform's information, which couldn't be done with a dict so you'd end up with a lot of ugly config[config['target_platform']['value'] lookups.

Reply:
if this has to be a class:
1. use ABCs indeed instead of a hand-written subset of dict methods
2. mark `update_config` etc as private - currently it looks like the users of config might just as well inject extra stuff at runtime
3. drop the class-global mutable attributes

Reply:
Good point, will update.